File: src/entities/Unit.ts
/**
 * Tipo runtime di un'unità (PG) sulla board.
 *
 * Stato baseline (vedi CLAUDE.md "Personaggio (baseline)"):
 *   HP 20, F/A/V 2/2/2, impeto 14, slancio 0, dadi azione 6
 *
 * L'unità occupa 7 esagoni (basetta) centrati su `position`. Vedi `getBaseHexes`.
 */

import { Axial } from '@core/hex/coords';
import { AcquiredSkill } from './Skill';

export type FactionId = 'A' | 'B';
export type UnitId = string;

/** ID di un pezzo di equipaggiamento (riferisce a `data/`) */
export type WeaponId = string;
export type ShieldId = string;
export type ArmorId = string;

export interface Unit {
  id: UnitId;
  name: string;
  faction: FactionId;

  /** ID del preset di provenienza (opzionale) — usato dall'AI matchup-aware per scegliere pesi appropriati */
  presetId?: string;

  /** Stat base (acquisto exp può modificarle? per ora baseline fissa 2/2/2) */
  forza: number;
  agilita: number;
  volonta: number;

  /** Punti vita */
  hp: number;
  hpMax: number;

  /** Impeto (iniziativa di base) */
  impeto: number;
  impetoMax: number;

  /** Slancio (modifica iniziativa, costa per movimento, si aggiunge a impeto) */
  slancio: number;

  /** Dadi azione disponibili e tetto massimo */
  dadiAzione: number;
  dadiAzioneMax: number;

  /** Equipaggiamento (riferimenti per ID a data/) */
  weapon?: WeaponId;
  /** Mano secondaria: scudo o seconda arma (per duello) */
  offhand?: WeaponId | ShieldId;
  armor?: ArmorId;

  /** Skill acquistate via character builder */
  skills: AcquiredSkill[];

  /** Posizione del centro della basetta sulla board */
  position: Axial;

  /** Stato vivente (true finché HP > 0) */
  alive: boolean;

  /** Numero di esagoni mossi nel turno corrente (1° gratis, gli altri costano 1 slancio).
   *  Reset a 0 a inizio turno. */
  hexMovedThisTurn: number;

  /** Stato di carica dell'arma equipaggiata. Rilevante per armi con `range.reload` (es. balestra).
   *  Default: true (carica all'inizio). Diventa false dopo un tiro. Si ricarica con azione RELOAD. */
  weaponLoaded: boolean;

  /** Flag "ho già usato la mia azione di turno" (attacco o ricarica).
   *  Le regole base permettono 1 azione per turno (movimento è separato). Reset a false in START_TURN. */
  actionTakenThisTurn: boolean;

  /** Fase 1 — Carica: posizione di inizio turno per calcolare delta_distance.
   *  Permette il calcolo di "quanti hex il PG si è avvicinato al nemico nel turno". */
  positionAtTurnStart?: Axial;

  /** Fase 1 — Posizione difensiva con scudo. Se true: parry.fixed scudo ×2 come RD passive
   *  (sia ranged sia CaC), imp scudo ×2, niente parry attivo. */
  defensiveStance: boolean;

  /** Fase 1 — Cap a 1 toggle/turno per la posizione difensiva. Reset in START_TURN. */
  defensiveToggledThisTurn: boolean;
}

/** Crea un'unità baseline ai valori di default. */
export function createBaselineUnit(params: {
  id: UnitId;
  name: string;
  faction: FactionId;
  position: Axial;
}): Unit {
  return {
    id: params.id,
    name: params.name,
    faction: params.faction,
    forza: 2,
    agilita: 2,
    volonta: 2,
    hp: 20,
    hpMax: 20,
    impeto: 14,
    impetoMax: 14,
    slancio: 0,
    dadiAzione: 6,
    dadiAzioneMax: 9, // pool max accumulabile (regola Valerio: 6 iniziali, 3/turno recovery, cap 9)
    skills: [],
    position: params.position,
    alive: true,
    hexMovedThisTurn: 0,
    defensiveStance: false,
    defensiveToggledThisTurn: false,
    weaponLoaded: true,
    actionTakenThisTurn: false,
  };
}

File: python/hex_tactics/entities/unit.py
"""
Tipo runtime di un'unità (PG) — port di src/entities/Unit.ts.

Stato baseline (CLAUDE.md "Personaggio (baseline)"):
  HP 20, F/A/V 2/2/2, impeto 14, slancio 0, dadi azione 6 (max 9)

L'unità occupa 7 esagoni (basetta) centrati su `position`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Literal, Optional

from hex_tactics.core.hex import Axial

from .skill import AcquiredSkill

FactionId = Literal["A", "B"]


@dataclass
class Unit:
    """Unità (PG) sulla board. Mutabile (lo state evolve via reducer)."""

    id: str
    name: str
    faction: FactionId

    # Stat base
    forza: int
    agilita: int
    volonta: int

    # HP
    hp: int
    hp_max: int

    # Iniziativa
    impeto: int
    impeto_max: int

    # Slancio (modifica iniziativa, costa per movimento)
    slancio: int

    # Dadi azione
    dadi_azione: int
    dadi_azione_max: int

    # Skill
    skills: List[AcquiredSkill]

    # Posizione centro basetta
    position: Axial

    # Stato vivente
    alive: bool

    # Movimento turno
    hex_moved_this_turn: int

    # Stato carica arma (rilevante per balestra ecc.)
    weapon_loaded: bool

    # Azione di turno già usata (1 per turno, movimento separato)
    action_taken_this_turn: bool

    # Equipaggiamento (ID che riferisce a data/)
    weapon: Optional[str] = None
    offhand: Optional[str] = None  # scudo o seconda arma
    armor: Optional[str] = None

    # 2026-05-05 NEW: inventario per armi da lancio single-use.
    # `thrown_inventory`: lista di armi extra da lancio (giavellotti, lance pronte).
    #   Quando si lancia weapon (throw single-use), si pop la prossima da qui.
    # `backup_weapon`: arma main alternativa estratta quando finiscono le throw.
    #   Es. lanciere con backup spada: lancia tutte le lance, poi estrae spada.
    thrown_inventory: List[str] = field(default_factory=list)
    backup_weapon: Optional[str] = None

    # ID preset di provenienza (per AI matchup-aware)
    preset_id: Optional[str] = None

    # M-9 (Fase 1): Carica — posizione di inizio turno per calcolare delta_distance
    # Permette di calcolare quanti hex il PG si è avvicinato al nemico durante il turno.
    # Resettato in apply_turn_start.
    position_at_turn_start: Optional[Axial] = None

    # M-10 (Fase 1): Posizione difensiva con scudo
    defensive_stance: bool = False  # se True: scudo offhand RD raddoppia, imp scudo raddoppia
    defensive_toggled_this_turn: bool = False  # max 1 toggle per turno

    # Numero di turni giocati dall'unità (incrementato in apply_turn_start).
    # Al primo turno (turns_played==0) NON c'è recovery dadi: il PG parte con 6 dadi
    # (pool iniziale baseline). Recovery scatta dal 2° turno in poi.
    turns_played: int = 0


def create_baseline_unit(
    *, id: str, name: str, faction: FactionId, position: Axial
) -> Unit:
    """Crea un'unità ai valori di default (HP 20, F/A/V 2/2/2, impeto 14, ...).

    Match con `createBaselineUnit` TS.
    """
    return Unit(
        id=id,
        name=name,
        faction=faction,
        forza=2,
        agilita=2,
        volonta=2,
        hp=20,
        hp_max=20,
        impeto=14,
        impeto_max=14,
        slancio=0,
        dadi_azione=6,
        dadi_azione_max=9,  # cap accumulabile (Valerio: 6 iniziali, 3/turno, cap 9)
        skills=[],
        position=position,
        alive=True,
        hex_moved_this_turn=0,
        weapon_loaded=True,
        action_taken_this_turn=False,
    )


__all__ = ["FactionId", "Unit", "create_baseline_unit"]
